lib/Dialect/LLHD/Transforms/LowerProcesses.cpp

```diff
@@ -0,0 +1,279 @@
+
+namespace circt {
+namespace llhd {
+#define GEN_PASS_DEF_LOWERPROCESSESPASS
+#include "circt/Dialect/LLHD/Transforms/LLHDPasses.h.inc"
+} // namespace llhd
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+using namespace llhd;
+using llvm::SmallDenseSet;
+using llvm::SmallSetVector;
+
+namespace {
+struct Lowering {
+  Lowering(ProcessOp processOp) : processOp(processOp) {}
+  void lower();
+  bool matchControlFlow();
+  void markObservedValues();
+  bool allOperandsObserved();
+  bool isObserved(Value value);
+
+  ProcessOp processOp;
+  WaitOp waitOp;
+  SmallDenseSet<Value> observedValues;
+};
+} // namespace
+
+void Lowering::lower() {
+  // Ensure that the process describes combinational logic.
+  if (!matchControlFlow())
+    return;
+  markObservedValues();
+  if (!allOperandsObserved())
+    return;
+  LLVM_DEBUG(llvm::dbgs() << "Lowering process " << processOp.getLoc() << "\n");
+
+  // Replace the process.
+  OpBuilder builder(processOp);
+  auto executeOp = builder.create<scf::ExecuteRegionOp>(
+      processOp.getLoc(), processOp.getResultTypes());
+  executeOp.getRegion().takeBody(processOp.getBody());
+  processOp.replaceAllUsesWith(executeOp);
+  processOp.erase();
+  processOp = {};
+
+  // Replace the `llhd.wait` with an `scf.yield`.
+  builder.setInsertionPoint(waitOp);
+  builder.create<scf::YieldOp>(waitOp.getLoc(), waitOp.getYieldOperands());
+  waitOp.erase();
+
+  // Simplify the execute op body region since disconnecting the control flow
+  // loop through the wait op has potentially created unreachable blocks.
+  IRRewriter rewriter(builder);
+  (void)simplifyRegions(rewriter, executeOp->getRegions());
+}
+
+/// Check that the process' entry block trivially joins a control flow loop
+/// immediately after the wait op.
+bool Lowering::matchControlFlow() {
+  // Ensure that there is only a single wait op in the process and that it has
+  // no destination operands.
+  for (auto &block : processOp.getBody()) {
+    if (auto op = dyn_cast<WaitOp>(block.getTerminator())) {
+      if (waitOp) {
+        LLVM_DEBUG(llvm::dbgs() << "Skipping process " << processOp.getLoc()
+                                << ": multiple wait ops\n");
+        return false;
+      }
+      waitOp = op;
+    }
+  }
+  if (!waitOp) {
+    LLVM_DEBUG(llvm::dbgs() << "Skipping process " << processOp.getLoc()
+                            << ": no wait op\n");
+    return false;
+  }
+  if (!waitOp.getDestOperands().empty()) {
+    LLVM_DEBUG(llvm::dbgs() << "Skipping process " << processOp.getLoc()
+                            << ": wait op has destination operands\n");
+    return false;
+  }
+
+  // Helper function to skip across empty blocks with only a single successor.
+  auto skipToMergePoint = [&](Block *block) -> std::pair<Block *, ValueRange> {
+    ValueRange operands;
+    while (auto branchOp = dyn_cast<cf::BranchOp>(block->getTerminator())) {
+      if (!block->without_terminator().empty())
```

Comment:
In the past I've frequently seen constants and `llhd.sig.extract` ops in the entry block. It should still be possible to do the lowering in that case. Or can this not occur anymore with your new mem2reg+deseq pipeline?

Reply:
I haven't seen constants and `llhd.sig.extract` in the entry block anymore. This might be due to the processes no longer being isolated from above. If you can pull an op into the entry block, chances are high that you can hoist it out of the process entirely. That might be the reason. But we should keep an eye out for these cases and add handling if they appear. My hope is that we can modify the process op (or create a separate flavor of processes) where you no longer have an entry block that you have to check for ops.